File: lib/dsp_blueprint_parser/md5f.rb

module DspBlueprintParser
  # class to generate DSP custom MD5F hash
  class Md5f
    @a
    @b
    @c
    @d

    @s11 = 7
    @s12 = 12
    @s13 = 17
    @s14 = 22
    @s21 = 5
    @s22 = 9
    @s23 = 14
    @s24 = 20
    @s31 = 4
    @s32 = 11
    @s33 = 16
    @s34 = 23
    @s41 = 6
    @s42 = 10
    @s43 = 15
    @s44 = 21

    def compute(message)
      array_to_hex_string(md5_array(message.force_encoding('UTF-8').each_byte.to_a), true)
    end

    private

    # @param x [Integer]
    # @param y [Integer]
    # @param z [Integer]
    # @return [Integer]
    def f(x, y, z)
      x & y | ~x & z
    end

    # @param x [Integer]
    # @param y [Integer]
    # @param z [Integer]
    # @return [Integer]
    def g(x, y, z)
      x & z | y & ~z
    end

    # @param x [Integer]
    # @param y [Integer]
    # @param z [Integer]
    # @return [Integer]
    def h(x, y, z)
      x ^ y ^ z
    end

    # @param x [Integer]
    # @param y [Integer]
    # @param z [Integer]
    # @return [Integer]
    def i(x, y, z)
      y ^ (x | ~z)
    end

    # @param a[Integer]
    # @param b [Integer]
    # @param c [Integer]
    # @param d [Integer]
    # @param mj [Integer]
    # @param s [Integer]
    # @param ti [Integer]
    def ff(a, b, c, d, mj, s, ti)
      a = a + f(b, c, d) + mj + ti
      a = a << s | a >> 32 - s
      a += b
      a
    end

    # @param a[Integer]
    # @param b [Integer]
    # @param c [Integer]
    # @param d [Integer]
    # @param mj [Integer]
    # @param s [Integer]
    # @param ti [Integer]
    def gg(a, b, c, d, mj, s, ti)
      a = a + g(b, c, d) + mj + ti
      a = a << s | a >> 32 - s
      a += b
      a
    end

    # @param a[Integer]
    # @param b [Integer]
    # @param c [Integer]
    # @param d [Integer]
    # @param mj [Integer]
    # @param s [Integer]
    # @param ti [Integer]
    def hh(a, b, c, d, mj, s, ti)
      a = a + h(b, c, d) + mj + ti
      a = a << s | a >> 32 - s
      a += b
      a
    end

    # @param a[Integer]
    # @param b [Integer]
    # @param c [Integer]
    # @param d [Integer]
    # @param mj [Integer]
    # @param s [Integer]
    # @param ti [Integer]
    def ii(a, b, c, d, mj, s, ti)
      a = a + i(b, c, d) + mj + ti
      a = a << s | a >> 32 - s
      a += b
      a
    end

    def md5_init
      @a = 1_732_584_193
      @b = 4_024_216_457
      @c = 2_562_383_102
      @d = 271_734_598
    end

    # @param input [Array<Integer>]
    # @return [Array<Integer>]
    def md5_append(input)
      num1 = 1
      length = input.size
      num2 = length % 64
      num3 = nil
      num4 = nil
      if num2 < 56
        num3 = 55 - num2
        num4 = length - num2 + 64
      elsif num2 == 56
        num3 = 63
        num1 = 1
        num4 = length + 8 + 64
      else
        num3 = 63 - num2 + 56
        num4 = length + 64 - num2 + 64
      end

      array_list = input.to_a
      array_list << 128 if num1 == 1

      index = 0
      while index < num3
        array_list << 0
        index += 1
      end

      num5 = length * 8
      num6 = num5 & 255
      num7 = num5 >> 8 & 255
      num8 = num5 >> 16 & 255
      num9 = num5 >> 24 & 255
      num10 = num5 >> 32 & 255
      num11 = num5 >> 40 & 255
      num12 = num5 >> 48 & 255
      num13 = num5 >> 56

      array_list << num6
      array_list << num7
      array_list << num8
      array_list << num9
      array_list << num10
      array_list << num11
      array_list << num12
      array_list << num13

      num_array = []
      index = 0
      while index * 4 < num4
        num_array << (array_list[index] || 0 | (array_list[index + 1] || 0 << 8) | (array_list[index + 2] || 0 << 16) | (array_list[index + 3] || 0 << 24))
        index += 1
      end

      num_array
    end

    # @param x [Array<Integer>]
    # @return [Array<Integer>]
    def md5_transform(x)
      puts x.size
      index = 0
      while index < x.size
        a = @a
        b = @b
        c = @c
        d = @d
        a = ff(a, b, c, d, x[index], 7, 3_614_090_360)
        d = ff(d, a, b, c, x[index + 1], 12, 3_906_451_286)
        c = ff(c, d, a, b, x[index + 2], 17, 606_105_819)
        b = ff(b, c, d, a, x[index + 3], 22, 3_250_441_966)
        a = ff(a, b, c, d, x[index + 4], 7, 4_118_548_399)
        d = ff(d, a, b, c, x[index + 5], 12, 1_200_080_426)
        c = ff(c, d, a, b, x[index + 6], 17, 2_821_735_971)
        b = ff(b, c, d, a, x[index + 7], 22, 4_249_261_313)
        a = ff(a, b, c, d, x[index + 8], 7, 1_770_035_416)
        d = ff(d, a, b, c, x[index + 9], 12, 2_336_552_879)
        c = ff(c, d, a, b, x[index + 10], 17, 4_294_925_233)
        b = ff(b, c, d, a, x[index + 11], 22, 2_304_563_134)
        a = ff(a, b, c, d, x[index + 12], 7, 1_805_586_722)
        d = ff(d, a, b, c, x[index + 13], 12, 4_254_626_195)
        c = ff(c, d, a, b, x[index + 14], 17, 2_792_965_006)
        b = ff(b, c, d, a, x[index + 15], 22, 968_099_873)
        a = gg(a, b, c, d, x[index + 1], 5, 4_129_170_786)
        d = gg(d, a, b, c, x[index + 6], 9, 3_225_465_664)
        c = gg(c, d, a, b, x[index + 11], 14, 643_717_713)
        b = gg(b, c, d, a, x[index], 20, 3_384_199_082)
        a = gg(a, b, c, d, x[index + 5], 5, 3_593_408_605)
        d = gg(d, a, b, c, x[index + 10], 9, 38_024_275)
        c = gg(c, d, a, b, x[index + 15], 14, 3_634_488_961)
        b = gg(b, c, d, a, x[index + 4], 20, 3_889_429_448)
        a = gg(a, b, c, d, x[index + 9], 5, 569_495_014)
        d = gg(d, a, b, c, x[index + 14], 9, 3_275_163_606)
        c = gg(c, d, a, b, x[index + 3], 14, 4_107_603_335)
        b = gg(b, c, d, a, x[index + 8], 20, 1_197_085_933)
        a = gg(a, b, c, d, x[index + 13], 5, 2_850_285_829)
        d = gg(d, a, b, c, x[index + 2], 9, 4_243_563_512)
        c = gg(c, d, a, b, x[index + 7], 14, 1_735_328_473)
        b = gg(b, c, d, a, x[index + 12], 20, 2_368_359_562)
        a = hh(a, b, c, d, x[index + 5], 4, 4_294_588_738)
        d = hh(d, a, b, c, x[index + 8], 11, 2_272_392_833)
        c = hh(c, d, a, b, x[index + 11], 16, 1_839_030_562)
        b = hh(b, c, d, a, x[index + 14], 23, 4_259_657_740)
        a = hh(a, b, c, d, x[index + 1], 4, 2_763_975_236)
        d = hh(d, a, b, c, x[index + 4], 11, 1_272_893_353)
        c = hh(c, d, a, b, x[index + 7], 16, 4_139_469_664)
        b = hh(b, c, d, a, x[index + 10], 23, 3_200_236_656)
        a = hh(a, b, c, d, x[index + 13], 4, 681_279_174)
        d = hh(d, a, b, c, x[index], 11, 3_936_430_074)
        c = hh(c, d, a, b, x[index + 3], 16, 3_572_445_317)
        b = hh(b, c, d, a, x[index + 6], 23, 76_029_189)
        a = hh(a, b, c, d, x[index + 9], 4, 3_654_602_809)
        d = hh(d, a, b, c, x[index + 12], 11, 3_873_151_461)
        c = hh(c, d, a, b, x[index + 15], 16, 530_742_520)
        b = hh(b, c, d, a, x[index + 2], 23, 3_299_628_645)
        a = ii(a, b, c, d, x[index], 6, 4_096_336_452)
        d = ii(d, a, b, c, x[index + 7], 10, 1_126_891_415)
        c = ii(c, d, a, b, x[index + 14], 15, 2_878_612_391)
        b = ii(b, c, d, a, x[index + 5], 21, 4_237_533_241)
        a = ii(a, b, c, d, x[index + 12], 6, 1_700_485_571)
        d = ii(d, a, b, c, x[index + 3], 10, 2_399_980_690)
        c = ii(c, d, a, b, x[index + 10], 15, 4_293_915_773)
        b = ii(b, c, d, a, x[index + 1], 21, 2_240_044_497)
        a = ii(a, b, c, d, x[index + 8], 6, 1_873_313_359)
        d = ii(d, a, b, c, x[index + 15], 10, 4_264_355_552)
        c = ii(c, d, a, b, x[index + 6], 15, 2_734_768_916)
        b = ii(b, c, d, a, x[index + 13], 21, 1_309_151_649)
        a = ii(a, b, c, d, x[index + 4], 6, 4_149_444_226)
        d = ii(d, a, b, c, x[index + 11], 10, 3_174_756_917)
        c = ii(c, d, a, b, x[index + 2], 15, 718_787_259)
        b = ii(b, c, d, a, x[index + 9], 21, 3_951_481_745)
        @a += a
        @b += b
        @c += c
        @d += d

        puts @a
        index += 16
        puts index
      end

      [@a, @b, @c, @d]
    end

    # @param input [Array<Integer>]
    # @return [Array<Integer>]
    def md5_array(input)
      md5_init
      num_array1 = md5_transform(md5_append(input))
      num_array2 = []

      binding.pry

      int index1 = 0
      while index1 < num_array1.size
        num_array2 << num_array1[index1] & 255
        num_array2 << num_array1[index1] >> 8 & 255
        num_array2 << num_array1[index1] >> 16 & 255
        num_array2 << num_array1[index1] >> 24 & 255
        index1 += 1
      end

      num_array2
    end

    # @param array [Array<Integer>]
    # @param uppercase [Boolean]
    # @return String
    def array_to_hex_string(array, uppercase)
      result = array.reduce('') { |accumulator, number| accumulator += number.to_s(16) }
      result = result.upcase if uppercase
      result
    end

    # @param input [Integer]
    # @param bits [Integer]
    # @return [Integer]
    def get_complement(input, bits)
      input
        .to_s(2)
        .rjust(bits, '0')
        .each_char
        .reduce('') { |accumulator, character| accumulator += character == '0' ? '1' : '0' }
        .to_i(2)
    end
  end
end

File: ext/md5f/md5f.c
#include<stdio.h> // define the header file  
#include <stdlib.h>
#include <string.h>
#include "ruby.h"
#include <extconf.h>

typedef struct _node {
  uint length;
  uint *array;
} node;

uint A;
uint B;
uint C;
uint D;

const int S11 = 7;
const int S12 = 12;
const int S13 = 17;
const int S14 = 22;
const int S21 = 5;
const int S22 = 9;
const int S23 = 14;
const int S24 = 20;
const int S31 = 4;
const int S32 = 11;
const int S33 = 16;
const int S34 = 23;
const int S41 = 6;
const int S42 = 10;
const int S43 = 15;
const int S44 = 21;

// ugly, I know
uint ArrayLength = 0;

uint F(uint x, uint y, uint z)
{
    return (uint) (((int) x & (int) y) | (~(int) x & (int) z));
}

uint G(uint x, uint y, uint z)
{
    return (uint) (((int) x & (int) z) | ((int) y & ~(int) z));
}

uint H(uint x, uint y, uint z)
{
    return x ^ y ^ z;
}

uint I(uint x, uint y, uint z)
{
    return y ^ (x | ~z);
}

uint FF(uint a, uint b, uint c, uint d, uint mj, int s, uint ti)
{
    a = a + F(b, c, d) + mj + ti;
    a = a << s | a >> (32 - s);
    a += b;
    return a;
}

uint GG(uint a, uint b, uint c, uint d, uint mj, int s, uint ti)
{
    a = a + G(b, c, d) + mj + ti;
    a = a << s | a >> (32 - s);
    a += b;
    return a;
}

uint HH(uint a, uint b, uint c, uint d, uint mj, int s, uint ti)
{
    a = a + H(b, c, d) + mj + ti;
    a = a << s | a >> (32 - s);
    a += b;
    return a;
}

uint II(uint a, uint b, uint c, uint d, uint mj, int s, uint ti)
{
    a = a + I(b, c, d) + mj + ti;
    a = a << s | a >> (32 - s);
    a += b;
    return a;
}

void MD5_Init()
{
    A = 1732584193U;
    B = 4024216457U;
    C = 2562383102U;
    D = 271734598U;
}

uint * MD5_Append(unsigned char input[], size_t length)
{
    int num1 = 1;
    // int length = input.Length;
    int num2 = length % 64;
    int num3;
    int num4 = 0;
    if (num2 < 56)
    {
        num3 = 55 - num2;
        num4 = length - num2 + 64;
    }
    else if (num2 == 56)
    {
        num3 = 63;
        num1 = 1;
        num4 = length + 8 + 64;
    }
    else
    {
        num3 = 63 - num2 + 56;
        num4 = length + 64 - num2 + 64;
    }

    unsigned char arrayList[num3 + 9 + length];
    uint position = 0;

    for (size_t i = 0; i < length; i++)
    {
        arrayList[position++] = input[i];
    }

    if (num1 == 1)
    {
        arrayList[position++] = (unsigned char) 128;
    }

    for (int index = 0; index < num3; ++index)
    {
        arrayList[position++] = (unsigned char) 0;
    }

    long num5 = (long) length * 8L;
    unsigned char num6 = (unsigned char) ((ulong) num5 & (ulong) 255);
    unsigned char num7 = (unsigned char) ((ulong) num5 >> 8 & (ulong) 255);
    unsigned char num8 = (unsigned char) ((ulong) num5 >> 16 & (ulong) 255);
    unsigned char num9 = (unsigned char) ((ulong) num5 >> 24 & (ulong) 255);
    unsigned char num10 = (unsigned char) ((ulong) num5 >> 32 & (ulong) 255);
    unsigned char num11 = (unsigned char) ((ulong) num5 >> 40 & (ulong) 255);
    unsigned char num12 = (unsigned char) ((ulong) num5 >> 48 & (ulong) 255);
    unsigned char num13 = (unsigned char) ((ulong) num5 >> 56);
    arrayList[position++] = num6;
    arrayList[position++] = num7;
    arrayList[position++] = num8;
    arrayList[position++] = num9;
    arrayList[position++] = num10;
    arrayList[position++] = num11;
    arrayList[position++] = num12;
    arrayList[position++] = num13;

    ArrayLength = num4 / 4;
    uint *numArray = calloc(ArrayLength, sizeof(uint));
    long index1 = 0;
    long index2 = 0;
    for (; index1 < (long) num4; index1 += 4)
    {
        numArray[index2] = (uint) ((int) arrayList[index1] | ((int) arrayList[index1 + 1] << 8) | ((int) arrayList[index1 + 2] << 16) | ((int) arrayList[index1 + 3] << 24));
        ++index2;
    }

    return numArray;
}

void MD5_Trasform(uint x[])
{
    for (int index = 0; index < ArrayLength; index += 16)
    {
        uint a = A;
        uint b = B;
        uint c = C;
        uint d = D;
        a = FF(a, b, c, d, x[index], 7, 3614090360U);
        d = FF(d, a, b, c, x[index + 1], 12, 3906451286U);
        c = FF(c, d, a, b, x[index + 2], 17, 606105819U);
        b = FF(b, c, d, a, x[index + 3], 22, 3250441966U);
        a = FF(a, b, c, d, x[index + 4], 7, 4118548399U);
        d = FF(d, a, b, c, x[index + 5], 12, 1200080426U);
        c = FF(c, d, a, b, x[index + 6], 17, 2821735971U);
        b = FF(b, c, d, a, x[index + 7], 22, 4249261313U);
        a = FF(a, b, c, d, x[index + 8], 7, 1770035416U);
        d = FF(d, a, b, c, x[index + 9], 12, 2336552879U);
        c = FF(c, d, a, b, x[index + 10], 17, 4294925233U);
        b = FF(b, c, d, a, x[index + 11], 22, 2304563134U);
        a = FF(a, b, c, d, x[index + 12], 7, 1805586722U);
        d = FF(d, a, b, c, x[index + 13], 12, 4254626195U);
        c = FF(c, d, a, b, x[index + 14], 17, 2792965006U);
        b = FF(b, c, d, a, x[index + 15], 22, 968099873U);
        a = GG(a, b, c, d, x[index + 1], 5, 4129170786U);
        d = GG(d, a, b, c, x[index + 6], 9, 3225465664U);
        c = GG(c, d, a, b, x[index + 11], 14, 643717713U);
        b = GG(b, c, d, a, x[index], 20, 3384199082U);
        a = GG(a, b, c, d, x[index + 5], 5, 3593408605U);
        d = GG(d, a, b, c, x[index + 10], 9, 38024275U);
        c = GG(c, d, a, b, x[index + 15], 14, 3634488961U);
        b = GG(b, c, d, a, x[index + 4], 20, 3889429448U);
        a = GG(a, b, c, d, x[index + 9], 5, 569495014U);
        d = GG(d, a, b, c, x[index + 14], 9, 3275163606U);
        c = GG(c, d, a, b, x[index + 3], 14, 4107603335U);
        b = GG(b, c, d, a, x[index + 8], 20, 1197085933U);
        a = GG(a, b, c, d, x[index + 13], 5, 2850285829U);
        d = GG(d, a, b, c, x[index + 2], 9, 4243563512U);
        c = GG(c, d, a, b, x[index + 7], 14, 1735328473U);
        b = GG(b, c, d, a, x[index + 12], 20, 2368359562U);
        a = HH(a, b, c, d, x[index + 5], 4, 4294588738U);
        d = HH(d, a, b, c, x[index + 8], 11, 2272392833U);
        c = HH(c, d, a, b, x[index + 11], 16, 1839030562U);
        b = HH(b, c, d, a, x[index + 14], 23, 4259657740U);
        a = HH(a, b, c, d, x[index + 1], 4, 2763975236U);
        d = HH(d, a, b, c, x[index + 4], 11, 1272893353U);
        c = HH(c, d, a, b, x[index + 7], 16, 4139469664U);
        b = HH(b, c, d, a, x[index + 10], 23, 3200236656U);
        a = HH(a, b, c, d, x[index + 13], 4, 681279174U);
        d = HH(d, a, b, c, x[index], 11, 3936430074U);
        c = HH(c, d, a, b, x[index + 3], 16, 3572445317U);
        b = HH(b, c, d, a, x[index + 6], 23, 76029189U);
        a = HH(a, b, c, d, x[index + 9], 4, 3654602809U);
        d = HH(d, a, b, c, x[index + 12], 11, 3873151461U);
        c = HH(c, d, a, b, x[index + 15], 16, 530742520U);
        b = HH(b, c, d, a, x[index + 2], 23, 3299628645U);
        a = II(a, b, c, d, x[index], 6, 4096336452U);
        d = II(d, a, b, c, x[index + 7], 10, 1126891415U);
        c = II(c, d, a, b, x[index + 14], 15, 2878612391U);
        b = II(b, c, d, a, x[index + 5], 21, 4237533241U);
        a = II(a, b, c, d, x[index + 12], 6, 1700485571U);
        d = II(d, a, b, c, x[index + 3], 10, 2399980690U);
        c = II(c, d, a, b, x[index + 10], 15, 4293915773U);
        b = II(b, c, d, a, x[index + 1], 21, 2240044497U);
        a = II(a, b, c, d, x[index + 8], 6, 1873313359U);
        d = II(d, a, b, c, x[index + 15], 10, 4264355552U);
        c = II(c, d, a, b, x[index + 6], 15, 2734768916U);
        b = II(b, c, d, a, x[index + 13], 21, 1309151649U);
        a = II(a, b, c, d, x[index + 4], 6, 4149444226U);
        d = II(d, a, b, c, x[index + 11], 10, 3174756917U);
        c = II(c, d, a, b, x[index + 2], 15, 718787259U);
        b = II(b, c, d, a, x[index + 9], 21, 3951481745U);
        A += a;
        B += b;
        C += c;
        D += d;
    }
}

unsigned char * MD5_Array(unsigned char input[])
{
    MD5_Init();
    uint *append = MD5_Append(input, strlen(input)); //todo length
    MD5_Trasform(append);
    uint numArray1[4] = { A, B, C, D};

    unsigned char *numArray2 = calloc(4 * 4, sizeof(unsigned char));

    int index1 = 0;
    int index2 = 0;
    while (index1 < 4)
    {
        numArray2[index2] = (unsigned char) (numArray1[index1] & (uint) 255);
        numArray2[index2 + 1] = (unsigned char) (numArray1[index1] >> 8 & (uint) 255);
        numArray2[index2 + 2] = (unsigned char) (numArray1[index1] >> 16 & (uint) 255);
        numArray2[index2 + 3] = (unsigned char) (numArray1[index1] >> 24 & (uint) 255);
        ++index1;
        index2 += 4;
    }

    free(append);
    return numArray2;
}

unsigned char * ArrayToHexString(unsigned char input[])
{
    size_t length = 16;
    unsigned char *result = calloc(length * 2, sizeof(unsigned char));

    for (size_t i = 0; i < length; i++)
    {
        sprintf(result+2*i, "%.2X", input[i]);
    }

    return result;
}

unsigned char * Compute(unsigned char message[])
{
    unsigned char * data = MD5_Array(message);
    unsigned char * result = ArrayToHexString(data);

    free(data);
    return result;
}

// int main()   // define the main function
// {
//     unsigned char input[18] = "very large string\0";
//     unsigned char *hex = Compute(input);
//     for (size_t i = 0; i < 32; i++)
//     {
//         printf("%C", hex[i]);
//     }

//     free(hex);
// }

VALUE MD5F = Qnil;  /* Ruby Module */

VALUE rb_print_length(VALUE self, VALUE str) {
    if (RB_TYPE_P(str, T_STRING) == 1) {
        return rb_sprintf("String length: %ld", RSTRING_LEN(str));
    }
    return Qnil;
}

void Init_foobar()
{
    MD5F = rb_define_module("MD5F");
    rb_define_method(MD5F, "print_length", rb_print_length, 1);
}